Make chips and add GluonCV training routine.

nholeman opened the pull request:
In this PR, 

For chips to be made, two methods are implemented: `process_scene_data` and `process_sceneset_results`. These methods give Raster Vision the relevant information to make chips from scenes.

In the `train` method, paths to the dataset and model are specified. These paths are passed to the `gluoncv_train` function, which runs the data through a GluonCV pipeline. Currently, this function only allows for `Resnet50v2` to be used. After each epoch's training, the model parameters are saved to a specified `model_params` uri.

Here are more details about what is used for training:

- **Net**: ResNet50v2, an uncontroversial model that balances accuracy and cost
- **Loss Function**: SoftmaxCrossEntropyLoss, which is cross entropy loss with the softmax function.
- **optimizer**: nag (Nesterov Accelerated Gradient), which calculates the gradient with approximated future parameters.
- **initialized weights**: Xavier.
- **metric**: F1, which balances precision and recall.

Below are the commands to run `train` locally:
```
export ROOT_URI=<local output dest>
rastervision -p plugin \
    run local -e spacenet.vegas \
    -a test True \
    -a use_remote_data False \
    -a root_uri ${ROOT_URI} \
    -a target buildings \
    -a task_type chip_classification
```

Below are the commands to run `train` remotely:
```
export ROOT_URI=<remote output dest>
rastervision -p plugin \
    run aws_batch -e spacenet.vegas \
    -a test True \
    -a use_remote_data True \
    -a root_uri ${ROOT_URI} \
    -a target buildings \
    -a task_type chip_classification
```

**Known Issues**: The batch size is currently quite small (32). This is because larger batch sizes resulted in an error when using the p2.xlarge's gpu: `mxnet.base.MXNetError: ... cudaMalloc failed: out of memory`.

lewfish commented on plugin/gluoncv_backend.py:
```diff
@@ -0,0 +1,433 @@
+            else:
+                io_def.add_input(conf.model_uri)
+
+        return (conf, io_def)
```
It looks like you are using an outdated commit of Raster Vision (some of the config builder stuff has changed since you started working on this). I think you should just continue developing against that outdated commit to avoid complicating things, but once everything works, then you can upgrade things.

lewfish commented:
Things to verify:
* Loss decreases locally and remotely
* Actually uses GPU when running remotely
* Writes model file to filesystem when running locally, and S3 when running remotely

lewfish commented:
> * **Loss Function**: SoftmaxCrossEntropyLoss, a linear classifier that uses cross entropy loss.

I don't think this is a "linear classifier". It measures how close the categorical distribution output by the softmax activation function is to the one hot encoding of the ground truth label.

lewfish commented on Dockerfile-gpu:
```diff
@@ -3,4 +3,11 @@ FROM raster-vision-gpu
+RUN pip install mxnet-cu90 --pre
+RUN pip install gluoncv --pre
+
+RUN sh -c "echo '/usr/local/cuda/lib64\n/usr/local/cuda/lib' >> /etc/ld.so.conf.d/nvidia.conf"
```
I'm assuming you copied this from some forum in order to fix a problem. In these situations it's best to include comment with a link to the original source which hopefully explains the rationale behind the incantation.

lewfish commented on plugin/train.py:
```diff
@@ -0,0 +1,136 @@
+from gluoncv.model_zoo import get_model
+
+
+def file_to_str(file_path):
```
RV already has a function called `file_to_str` :) It's in `rastervision.utils.files`

nholeman replied:
Function deleted; it was not being used anyway.

lewfish commented on plugin/train.py:
```diff
@@ -0,0 +1,136 @@
+    return chip
+
+
+def model_build(classes, pretrained_model_options=None):
```
It's more typical for function names to start with the verb (ie. `build_model`)

lewfish replied:
If `classes` in the number of classes, a clearer name would be `num_classes`

lewfish replied:
`pretrained_model_options` is not used anywhere in this function.

lewfish commented on plugin/train.py:
```diff
@@ -0,0 +1,136 @@
+
+def model_build(classes, pretrained_model_options=None):
+    num_cpus = 1
+    ctx = [mx.cpu(i) for i in range(num_cpus)]
```
Since the context was created in the `gluoncv_train` function and is passed into `test`, it seems like it should also be passed into this function instead of re-creating it here.

nholeman replied:
The reason for the redundancy is that the `load_model` method in `gluoncv_backend.py` will use `model_build()`, so `ctx` will have to be created twice at some point. I wanted to isolate as much mxnet-specific code to `train.py` as possible, as opposed to importing mxnet to `gluoncv_backend.py` as well.

lewfish commented on plugin/train.py:
```diff
@@ -0,0 +1,136 @@
+                                          batch_axis=0)
+        label = gluon.utils.split_and_load(batch[1], ctx_list=ctx,
+                                           batch_axis=0)
+        with ag.record():
```
It seems strange that you need to use autograd just for making predictions on the test set. You should add a comment that you added it to make things work and don't know why it helps (unless you now know why it's needed in which case you should provide the reason).

lewfish replied:
I haven't read this yet, but it seems relevant: https://mxnet.incubator.apache.org/api/python/autograd/autograd.html#train-mode-and-predict-mode

lewfish commented on plugin/train.py:
```diff
@@ -0,0 +1,136 @@
+                         files, including classes and training_data_dir.
+    arg str model_params: path to pretrained model
+
+    return: trained model
```
We're following a different convention for PyDocs (see any source code in RV for example) so this should be updated.

lewfish replied:
Since this was adapted from the GluonCV examples you should note this and provide a link back to the original code in case people want to see what you changed, etc.

lewfish commented on plugin/train.py:
```diff
@@ -0,0 +1,136 @@
+    lr_decay_epoch = [10, 20, 30, 40, np.inf]
+
+    num_cpus = 1
+    ctx = [mx.cpu(i) for i in range(num_cpus)]
```
What's the mechanism for getting this to work on GPUs? I was expecting to see an `mx.gpu()` somewhere.

nholeman replied:
You are correct. `mx.cpu()` must be manually changed to `mx.gpu()` in order for this to work on GPUs. I will specify this in the PR.

lewfish commented on plugin/train.py:
```diff
@@ -0,0 +1,136 @@
+    num_cpus = 1
+    ctx = [mx.cpu(i) for i in range(num_cpus)]
+    # GluonCV Model build here
+    net = get_model('ResNet50_v2', classes=classes, ctx=ctx)
```
There should be an option for using a pretrained model or not, but as we discussed we're not worrying about adding various options at this time. But as a default, we should be using a pretrained model since that almost always improves performance by a lot. So you should use the pretrained=True flag here. https://gluon-cv.mxnet.io/api/model_zoo.html#gluoncv.model_zoo.get_model

nholeman replied:
The pretrained `ResNet50v2` expects `classes=1000`, whereas in the Vegas buildings example, `classes=2`. This results in an incompatible shape.

nholeman replied:
Fixed. You can set pretrained to `True`, then change the structure of the output layer to match the number of classes.

lewfish commented on plugin/train.py:
```diff
@@ -0,0 +1,136 @@
+              'Val-f1: %.3f | time: %.1f' %
+              (epoch, train_f1, train_loss, val_f1, time.time() - tic))
+
+        model.save_parameters(model_params)
```
The PyDocs says that `model_params` is the path to the pretrained model weights, but that's not how you're using that variable here. A better name would be `model_path` and then just update the comment.

lewfish replied:
You're only saving the model at the very end, so that means the periodic sync has nothing to upload. Getting checkpointing to work is going to take some time that we don't have right now, so I would just add a TODO comment about that.

nholeman replied:
`model.save_parameters()` is within the `for epoch in range(epochs)` loop, meaning that the model is saved at each epoch. Should it be saved after each batch instead?

lewfish commented on plugin/gluoncv_backend.py:
```diff
@@ -0,0 +1,439 @@
+                             checkpoing if exists, if the backend allows
+                             for this.
+        """
+        b = deepcopy(self)
```
~~You should get rid of this method since none of these options are actually implemented. It could be confusing otherwise.~~ Since the work is already done to get these options into the gluon_cv training routine we should just leave it as is, but raise an exception that says these option aren't implemented yet or something. I would normally say to remove dead code, but it seems like this will be useful in the future and I don't want to waste time adding the stuff back at that time.

lewfish commented on plugin/gluoncv_backend.py:
```diff
@@ -0,0 +1,439 @@
+                    config_mod,
+                    ignore_missing_keys=False,
+                    set_missing_keys=False):
+        """Given a dict, modify the tensorflow pipeline configuration
```
Comment refers to Tensorflow.

lewfish commented on plugin/gluoncv_backend.py:
```diff
@@ -0,0 +1,439 @@
+        return (config, pretrained_model_path)
+
+
+class GluonCVBackend(Backend):
```
I would move the actual backend and associated functions into a separate file to mimic the way the other backends are implemented (ie. separate `<backend>_config.py` and `<backend>.py` files)

lewfish reviewed (changes requested):
I still need to review the `GluonCVBackend` class but this should be enough to start revising.

lewfish commented on plugin/train.py:
```diff
@@ -0,0 +1,199 @@
+    Args:
+        config_path: default dict containing relevant configuration
+                     files, including classes and training_data_dir.
+        model_path: path to pretrained model
```
Comment is inaccurate - It's a path to the model that is going to be trained

lewfish commented on plugin/train.py:
```diff
@@ -0,0 +1,199 @@
+    net.output.initialize(init.Xavier(), ctx=ctx)
+
+    net.collect_params().reset_ctx(ctx)
+    net.hybridize()
```
A comment explaining why these two lines are needed would be useful, or a link to code that you copied if that's the case.

lewfish commented on plugin/gluoncv_config.py:
```diff
@@ -0,0 +1,233 @@
+           such that keys that are found recursively in the configuration
+           are replaced with those values. TODO: better explination.
+        """
+        # TODO exlain following line
```
Should fill in TODO and I think this should also be in the with_train_options since none of them are being used.

lewfish reviewed (changes requested):
Also, in the PR description the commands to run are inaccurate. For running locally, it should be the following 
```
rastervision -p plugin run local \
      -e spacenet.vegas      -a test True      -a use_remote_data False  \
      -a root_uri ${ROOT_URI}      -a target buildings      -a task_type chip_classification
```

nholeman committed:
Chip and Train commands succeed

Chip and train commands succeed locally and remotely on the Vegas Spacenet dataset. See PR for details.

lewfish commented:
Ready for you to merge.
